File: problems/p673/p673.go
package p673

func findNumberOfLIS(nums []int) int {
	var largest int
	var cntMap = make([]int, len(nums))
	var dp = make([]int, len(nums))
	for i := range dp {
		dp[i] = 1
		cntMap[i] = 1
	}
	for i := 0; i < len(nums); i++ {
		for j := 0; j < i; j++ {
			if nums[i] > nums[j] && dp[i] < dp[j]+1 {
				dp[i] = dp[j] + 1
				cntMap[i] = cntMap[j]
			} else if nums[i] > nums[j] && dp[i] == dp[j]+1 {
				cntMap[i] += cntMap[j]
			}
		}
		largest = max(largest, dp[i])
	}
	var result int
	for i := range dp {
		if dp[i] == largest {
			result += cntMap[i]
		}
	}

	return result
}

func max(i, j int) int {
	if i > j {
		return i
	}
	return j
}

File: problems/p673/p673.py
from typing import List


class Solution:
    def findNumberOfLIS(self, nums: List[int]) -> int:
        largest = 0
        dp = [1]*len(nums)
        cnt_map = {}
        for i in range(0, len(nums)):
            cnt_map[i] = 1
        for i in range(len(nums)):
            for j in range(i):
                if nums[i] > nums[j] and dp[i] < dp[j]+1:
                    dp[i] = dp[j]+1
                    cnt_map[i] = cnt_map[j]
                elif nums[i] > nums[j] and dp[i] == dp[j]+1:
                    cnt_map[i] += cnt_map[j]
            largest = max(largest, dp[i])
        result = 0
        for i in range(len(nums)):
            if dp[i] == largest:
                result += cnt_map[i]

        return result
